File: src/main/java/com/example/visainterview/entity/TransactionType.java
package com.example.visainterview.entity;

import lombok.Getter;

@Getter
public enum TransactionType {
    NORMAL_PURCHASE(1),
    PURCHASE_WITH_INSTALLMENTS(2),
    WITHDRAWAL(3),
    CREDIT_VOUCHER(4);

    private final int code;

    TransactionType(int code) {
        this.code = code;
    }

    public static TransactionType fromCode(int code) {
        for (TransactionType type : values()) {
            if (type.code == code) {
                return type;
            }
        }
        throw new IllegalArgumentException("Invalid transaction type code: " + code);
    }

    public boolean isDebit() {
        return this == NORMAL_PURCHASE || this == WITHDRAWAL;
    }

    public boolean isCredit() {
        return this == CREDIT_VOUCHER;
    }
}

File: golang/models/transaction_type.go
package models

type TransactionType int

const (
	NormalPurchase TransactionType = 1
	PurchaseWithInstallments TransactionType = 2
	Withdrawal TransactionType = 3
	CreditVoucher TransactionType = 4
)

func (t TransactionType) IsValid() bool {
	return t >= 1 && t <= 4
}

func (t TransactionType) IsDebit() bool {
	return t == NormalPurchase || t == Withdrawal
}

func (t TransactionType) IsCredit() bool {
	return t == CreditVoucher
}

func (t TransactionType) String() string {
	switch t {
	case NormalPurchase:
		return "normal_purchase"
	case PurchaseWithInstallments:
		return "purchase_installments"
	case Withdrawal:
		return "withdrawal"
	case CreditVoucher:
		return "credit_voucher"
	default:
		return "unknown"
	}
}
